### ISSUE TITLE
Different results between python and C# for the same onnx/directml runtime

### ISSUE THREAD

**Issue by VivitionDeveloper** [microsoft/onnxruntime#27402]
After struggling a day, CoPilot created a readme md file containing all information. 

In short: processing the same input image, using the same ONXX Runtime (version 1.23) on the same system causes completely different results between C# and Python. See below for details. All required files for reproduction have been attached.
- the input image (is also available at: https://github.com/XuJiacong/PIDNet/tree/main/samples)
<img width="2048" height="1024" alt="Image" src="https://github.com/user-attachments/assets/ea455691-ce55-444b-9227-4ceb907a9288" />
- the expected result (is also available at: https://github.com/XuJiacong/PIDNet/tree/main/samples/out). This is a colorized mask.
<img width="2048" height="1024" alt="Image" src="https://github.com/user-attachments/assets/fbffad8a-929d-424f-b4da-35bb16a8fa16" />
- the used data model, however, I was not able to attach that, as it was too large (ca 130MB). Please let me know how to provide that file.

# ImageNet Normalization Produces Different Output in C# vs Python

## Summary
When applying ImageNet normalization (`mean=[0.485, 0.456, 0.406]`, `std=[0.229, 0.224, 0.225]`) to input tensors, ONNX Runtime C# produces significantly different inference results compared to Python, despite identical input data and model.

## Environment
- **ONNX Runtime Version**: 1.23.0
- **Platform**: Windows 11, x64
- **Model**: PIDNet-L (semantic segmentation, 19 classes)
- **Input Shape**: `[1, 3, 1024, 2048]` (NCHW format)
- **Execution Provider**: CPU (CPUExecutionProvider)

### Package Versions
**Python:**
```
onnxruntime-directml==1.23.0
numpy==1.26.4
```

**C#:**
```xml
<PackageVersion Include="Microsoft.ML.OnnxRuntime" Version="1.23.0" />
<PackageVersion Include="Microsoft.ML.OnnxRuntime.DirectML" Version="1.23.0" />
```

### Input Tensor Verification - COMPLETE STATISTICS
Both Python and C# produce **IDENTICAL** input tensors (within floating point precision):

**Global Tensor Statistics (after ImageNet normalization):**
```
Python: Mean=-0.4636039436, Std=0.9593937397, Min=-2.1179039478, Max=2.6399998665
C#:     Mean=-0.4636040879, Std=0.9593940572, Min=-2.1179039478, Max=2.6400001049
        (difference: ±0.0000003 = floating point rounding error, IDENTICAL)
```

**First 10 flattened tensor values:**
```
Python: [-2.117904, -2.0665295, -1.9466563, -1.8439078, -1.7411593, -1.6555356, -1.5699117, -1.5014127, -1.415789, -1.34729]
C#:     [-2.117904, -2.066530,  -1.946656,  -1.843908,  -1.741159,  -1.655536,  -1.569912,  -1.501413,  -1.415789, -1.347290]
```

### Output Tensor Discrepancy
Despite **mathematically identical inputs**, the **output logits are completely different**:

**Center pixel logits (19 classes):**
```
Python (class 0 = max):  [9.081, 3.707, 0.169, -0.699, -1.273, -1.710, -2.155, -1.033, ...]
C# (class 8 = max):      [-5.139, -6.596, 4.358, -7.223, 5.573, 4.321, -1.109, 7.800, 9.110, ...]
```

**Result similarity:**
```
Python → Output image: 99.99% match with expected ground truth
C# → Output image:     6.84% match with expected ground truth
```

### Python Code (Working - 99.99% similarity)
```python
import numpy as np
import cv2
import onnxruntime

mean = [0.485, 0.456, 0.406]
std = [0.229, 0.224, 0.225]

img = cv2.imread("test.png", cv2.IMREAD_COLOR)
img = cv2.resize(img, (2048, 1024))
img = img.astype(np.float32)[:, :, ::-1]  # BGR→RGB
img = img / 255.0
img -= mean
img /= std
img = img.transpose((2, 0, 1)).copy()  # HWC→CHW
img = np.expand_dims(img, axis=0)  # Add batch

session = onnxruntime.InferenceSession("model.onnx", providers=['CPUExecutionProvider'])
output = session.run(None, {"image": img})[0]
# Result: 99.99% match with expected output
```

### C# Code (Failing - 6.84% similarity)
```csharp
using OpenCvSharp;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;

var bgrImage = Cv2.ImRead("test.png", ImreadModes.Color);
using var resized = new Mat();
Cv2.Resize(bgrImage, resized, new Size(2048, 1024));

using var rgb = new Mat();
Cv2.CvtColor(resized, rgb, ColorConversionCodes.BGR2RGB);

using var floatMat = new Mat();
rgb.ConvertTo(floatMat, MatType.CV_32FC3, 1.0 / 255.0);

float[] mean = [0.485f, 0.456f, 0.406f];
float[] std = [0.229f, 0.224f, 0.225f];

var tensor = new DenseTensor<float>([1, 3, 1024, 2048]);

unsafe {
    float* ptr = (float*)floatMat.Data;
    for (int c = 0; c < 3; c++) {
        for (int y = 0; y < 1024; y++) {
            for (int x = 0; x < 2048; x++) {
                int srcIdx = (y * 2048 + x) * 3;
                tensor[0, c, y, x] = (ptr[srcIdx + c] - mean[c]) / std[c];
            }
        }
    }
}

var session = new InferenceSession("model.onnx");
var inputs = new List<NamedOnnxValue> { NamedOnnxValue.CreateFromTensor("image", tensor) };
var results = session.Run(inputs);
// Result: Only 6.84% match with expected output
```

### Workaround
**Removing normalization** (simple `[0-1]` scaling) restores similarity:
```csharp
tensor[0, c, y, x] = ptr[srcIdx + c];  // Just [0-1], no mean/std
```
**Result:** 92-93% similarity (acceptable, but not ideal)

## Tested Variations (All Failed)
1. ✅ **OrtValue API instead of DenseTensor** - No difference (still 6.84%)
2. ✅ **Explicit SessionOptions matching Python** - No difference
3. ✅ **Pre-allocated float array** - No difference
4. ✅ **Multiple tensor construction methods** - No difference
5. ✅ **ONNX model verification** - Models are byte-identical
6. ✅ **Image data verification** - Images are byte-identical
7. ✅ **Complete tensor statistics validation** - Tensors are mathematically identical (±0.0000003)

**All variations produce 6.84% similarity - proving the issue is in ONNX Runtime C# core, not tensor construction.**

## Expected Behavior
C# ONNX Runtime should produce the same output as Python when given identical input tensors and model.

## Actual Behavior
C# produces completely different logits, suggesting the tensor data is being interpreted differently internally despite:
- Identical input tensor values
- Identical ONNX model file
- Same ONNX Runtime version (1.23.0)
- Same execution provider (CPU)

## Additional Notes
- The issue is **specific to normalization** - without it, C# achieves 90-93% similarity
- This suggests a potential bug in how C# bindings handle normalized float data
- The model works perfectly in Python (99.99% match with expected output)

## Request
Please investigate why ONNX Runtime C# produces different inference results when using normalized input data compared to Python bindings with the same runtime version.

**Comment by VivitionDeveloper** [comment]
May be, the previous thing is false alarm. I'v changed the pre-processing and post-processing steps and now get the same output. Causes were: different resize algorithms between PyTorch and C# opencv. Rounding issues (float vs double) and some more of these quirky things.
